pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala

```diff
@@ -0,0 +1,77 @@
+package uk.ac.wellcome.platform.merger.rules
+import uk.ac.wellcome.models.work.internal.{
+  TransformedBaseWork,
+  UnidentifiedWork
+}
+import uk.ac.wellcome.platform.merger.rules.WorkFilters.WorkFilter
+
+case class MergeResult[T](field: T, redirects: Seq[TransformedBaseWork])
+
+/*
+ * A trait to extend in order to merge fields of the type member `Field`
+ *
+ * The implementor must provide the method `merge`, which takes a target
+ * work and a list of source works, and returns the new value of the field
+ * as well as a list of source works that this rule would like to be redirected.
+ *
+ * Many merge rules will want to apply different logic given conditions on
+ * both the target and the source works.
+ * The PartialRule trait is provided to achieve this: in addition to
+ * a `rule` that has the same signature as the `merge` method, the implementor
+ * must provide predicates for valid targets and sources.
+ *
+ * Because PartialRule is a PartialFunction, the `rule` will never be called
+ * for targets/sources that don't satisfy these predicates, and we gain things
+ * like `orElse` and `andThen` for free.
+ */
+trait FieldMergeRule {
+  protected final type Params = (UnidentifiedWork, Seq[TransformedBaseWork])
+  protected type Field
+
+  def merge(target: UnidentifiedWork,
+            sources: Seq[TransformedBaseWork]): MergeResult[Field]
+
+  protected val identityOnTarget: PartialFunction[Params, UnidentifiedWork] = {
+    case (target, _) => target
+  }
+
+  protected trait PartialRule extends PartialFunction[Params, Field] {
+    val isDefinedForTarget: WorkFilter
+    val isDefinedForSource: WorkFilter
+
+    def rule(target: UnidentifiedWork, sources: Seq[TransformedBaseWork]): Field
+
+    override def apply(x: Params): Field =
+      x match {
+        case (target, sources) =>
+          rule(target, sources.filter(isDefinedForSource))
+      }
+
+    override def isDefinedAt(x: Params): Boolean = x match {
+      case (target, sources) =>
+        isDefinedForTarget(target) && sources.exists(isDefinedForSource)
+    }
+  }
+}
+
+/*
+ * PartialRules can be used with `orElse` as a chain of fallbacks for picking a
+ * field off sources, or they can be composed.
+ *
+ * Extending the ComposedFieldMergeRule trait provides a helper, `composeRules`,
+ * which takes a list of rules, as well as a helper function to copy the new
+ * merged Fields from each step into the "running-total" target, and composes
+ * them right-to-left.
+ */
+trait ComposedFieldMergeRule extends FieldMergeRule {
+  protected final def composeRules(
```

Comment:
I don't personally see the need for this to exist in a separate trait, as could just be included in the base one and only be used when required. I guess it could be argued it is more explicit, but it's at the cost of a more complex type hierarchy.